### ISSUE TITLE
Fonts!

### ISSUE THREAD

**Issue by asinghvi17** [JuliaPlots/MakieTeX.jl#36]
Some useful resources for font changes:

- Known good math fonts https://tex.stackexchange.com/questions/425098/which-opentype-math-fonts-are-available
- Use mathspec for non-known good fonts
- Allow the user to specify the math font, otherwise default to lmodern
- Create a flexible system to use these fonts
- (Possibly) ship or at least link to the fonts?

**Comment by Moelf** [comment]
from trying the package it's clear that `font = blah"` doesn't work with MakieTeX.jl right now.

What's needed for the interaction to happen? I though as a naive implementation we just need to inject something like this into the tectonic input right?
```
\usepackage{fontspec}%
\newfontfeature{Microtype}{protrusion=default;expansion=default;}%
\setmainfont{texgyrepagella-regular.otf}[%
  Microtype,
  Ligatures = TeX,
  BoldFont = texgyrepagella-bold.otf,
  ItalicFont = texgyrepagella-italic.otf,
  BoldItalicFont = texgyrepagella-bolditalic.otf,
]%
```

**Comment by asinghvi17** [comment]
Something like that, but we would need to figure out how to get all the other font types as well as a good math font.

Also, the way the text algorithm works now precludes MakieTeX from hooking into it as it previously did.  I'd have to refactor that algorithm to (preferably) accept scatters as well, which could then be used for both emojis and images.

**Comment by Moelf** [comment]
ok rn I just want a minimal working thing so I can be compliant with publication requirement with Makie and the only requirement is to render something like 
```
\sqrt{s} = 13 \text{TeV}
```
using Helvetica (or Nimbus).

- - -

I think for Makie's use case, i.e., `text!(L"....", font = )`, it's fine to replace everything (mainfont, mathfont ... ) in .tex file, my reasoning is that, users are most likely just adding annotation or axis labels, not trying to write an essay with MakieTeX, we don't need a lot of fancy automatic detection fall back select best blah blah, just do what user wants as a start

**Comment by Moelf** [comment]
> the way the text algorithm works now precludes MakieTeX from hooking into it as it previously did.

are you saying that right now there's no method in this repo that can receive `font = blah` from Makie?

**Comment by asinghvi17** [comment]
yes, even the old `test!(L"...")` uses MathTeXEngine as it does in base Makie

**Comment by asinghvi17** [comment]
> requirement is to render something like

How would you do that without a Helvetica math font anyway?  Unless I misunderstand this would show up in CMU even in normal latex?

**Comment by Moelf** [comment]
I don't know but matplotlib (which goes through `texlive`) can render it: https://gitlab.sauerburger.com/cern/fsauerbu/atlasify/-/blob/master/atlasify/__init__.py#L60

in any case, when using `Helvetica`, `\sqrt{s}` looks very different compared to default font CMU

**Comment by Moelf** [comment]
>yes, even the old test!(L"...") uses MathTeXEngine as it does in base Makie

wait, so is this pkg right now completely useless? what does it even do if it can't intercept `text!()` call?

**Comment by asinghvi17** [comment]
provides an LTeX layoutable which can render arbitrary latex, that's about it

**Comment by asinghvi17** [comment]
for now, if you know which math font you want you could provide a custom preamble to LTeX, like so:

```julia
tex_to_render = TeXDocument("this is a full and compilable latex document pasted into a string in Julia")
LTeX(fig[i, j]; tex = tex_to_render, ...)
```

and then you can load your custom fonts etc.

**Comment by Moelf** [comment]
I'd be happy to help since I know a little bit around Makie code base by now

**Comment by asinghvi17** [comment]
Am testing out the text refactor-ish thing now...will allow an extension to be added for Pango pretty easily as well.

**Comment by Moelf** [comment]
ok here's a super hacky idea:
1. make a function here, something like `textext!(fig, Lstring; fonts, color, size... <whatever we want>)`
2. this function does a few things
    - implement the injection of fonts and color as we see fit (basically extend what this repo already does)
    - render the `tex` text to a image (not sure what we're doing pdf or png)
    - call `Makie.scatter!(...)` with `marker = load(image_path.png)`
3. since we `load()`ed the image, we have chance to pass the correct `size` tuple to `scatter!()` so the image marker has the correct aspect ratio

**Comment by asinghvi17** [comment]
We mostly have that in `TeXLabel`, it just needs the font stuff implemented.  Backgroundcolor, color, etc is all implemented already.

**Comment by asinghvi17** [comment]
If you have an idea of how we can: 
(a) figure out how to get fonts into LaTeX generally, given a font family and a path to its file, 
(b) figure out a good correlation between math fonts and user fonts (maybe a lookup table which someone has already implemented in the LaTeX world?)

then I think those would be the natural next steps.

**Comment by asinghvi17** [comment]
Ok, that sounds good.  Yeah I guess we can assume that any font a user provides must be installed, I'm not the most familiar with fontconfig but it seems you can specify bold/italic/etc fonts as separate font files.  The way I see this, there are two cases:

1.  Makie doesn't really support this, but if all fonts are in one file, then how would we support this?
2. If all fonts are in separate files, this becomes pretty straightforward.

I guess `fonts[:regular]` would map to `mainfont`, bold to `boldfont`, italic to `italicfont`, etc.  Most fonts describe their styles in a space-separated way - can we change that to either underscored or pascal cased (first letter of each word is capitalized)?

We could just assume a CMU math font until told otherwise by the user, that works.

**Comment by Moelf** [comment]
```julia
julia> function main()
           tex = CachedTeX(MakieTeX.texdoc(raw"$\sqrt{s} = 13 \text{TeV}$"));
           f,a,_ = plot(1:3, 1:3)
           marker = MakieTeX.recordsurf2img(tex)
           scatter!(a, 1.5, 1.5; marker, markersize=100)
           CairoMakie.save("/tmp/jl_onm7cZSLiJ.png", f)
       end
```

![image](https://user-images.githubusercontent.com/5306213/216679129-a0e0690c-5c6d-43b9-a2e0-511e86359e0b.png)


- - -
I don't understand why it's rotated (I thought `recordsurf2img` already corrected for it) and also why the aspect ratio is wrong

**Comment by Moelf** [comment]
ok, if we apply `rotl90` and also use `CachedTex.dims`, we can get desired result
